zerver/lib/actions.py

```diff
@@ -4803,6 +4803,7 @@ def do_change_default_sending_stream(
         )
 
 
+@transaction.atomic
```

Comment:
Similarly here, why don't we put this only around a part of the function, that deals with the db operation, and then we won't have to worry about adding `on_commit` etc. in other places?
```
    user_profile.save(update_fields=["default_events_register_stream"])
    event_time = timezone_now()
    RealmAuditLog.objects.create(
        realm=user_profile.realm,
        event_type=RealmAuditLog.USER_DEFAULT_REGISTER_STREAM_CHANGED,
        event_time=event_time,
        modified_user=user_profile,
        acting_user=acting_user,
        extra_data=orjson.dumps(
            {
                RealmAuditLog.OLD_VALUE: old_value,
                RealmAuditLog.NEW_VALUE: None if stream is None else stream.id,
            }
        ).decode(),
    )
```

Reply:
I am not sure if it was completely decided but [this] conversation seems to point out that we are fine with using `on_commit` for events.

Reply:
I think the result will be the same with `transaction.on_commit` to send the event as with sending the event just after the end of the translation block.  In both cases, we're vulnerable to a race where the process fails (E.g. is killed by an external event) just after completing the database transaction but before the event is sent.

So it's really a question of which provides better readability. This is probably a question worth starting a thread on #backend about for visibility of discussion.